### PR TITLE
Check dynamic mode before applicability. Do not use non-dymanic or slow/narrow Winograds when cutting corners.

### PR DESCRIPTION
**Pull request by JehandadKhan**
- Accelerates DYNAMIC_HYBRID Find mode by checking `IsDynamic()` prior `IsApplicable()`.
  - `IsApplicable()` is much longer, so when solver is `!IsDynamic()` we are wasting time if applicability is checked prior dynamicity.
- Accelerates all Find modes by using only dynamic and "fast-and-wide-enough" Winograd solvers in the `is_winograd_only` hackery in Find.

### CODE REVIEW

File: src/include/miopen/find_solution.hpp [hunk 1]
```diff
@@ -151,10 +151,10 @@ struct SolverContainer
                 if(find_only.IsValid() && find_only != Id{SolverDbId(solver)})
                 { // Do nothing (and keep silence for the sake of Tuna), just skip.
                 }
-                else if(!solver.IsApplicable(search_params))
-                    MIOPEN_LOG_I2(SolverDbId(solver) << ": Not applicable");
                 else if(search_params.use_dynamic_solutions_only && !solver.IsDynamic())
                     MIOPEN_LOG_I2(SolverDbId(solver) << ": Skipped (non-dynamic)");
+                else if(!solver.IsApplicable(search_params))
+                    MIOPEN_LOG_I2(SolverDbId(solver) << ": Not applicable");
```
Comment (atamazov) [inline src/include/miopen/find_solution.hpp:157]: Optimization for speed?

**Comment by atamazov** [comment]
@JehandadKhan I feel that you found some issue, possibly potential... Could you please describe the use case? Most likely some additional fixes are needed.

**Comment by atamazov** [comment]
Perhaps the best solution would be: if `IsWinograd3x3SupportedAndFast()`, then narrow the set of Winograd solvers to single `ConvBinWinograd3x3U` everywhere.

File: src/include/miopen/find_solution.hpp (hunk 1 above)
Comment (JehandadKhan) [reply src/include/miopen/find_solution.hpp:157]: yes, some `IsApplicable` methods can take as long as 200ms which causes considerable overhead.

**Comment by atamazov** [comment]
@JehandadKhan This should be re-implemented in accordance to https://github.com/ROCmSoftwarePlatform/MIOpen/pull/544#issuecomment-715607901. Please let me know if you agree, and I will do that right in this branch.

**Comment by JehandadKhan** [comment]
> 
> 
> @JehandadKhan This should be re-implemented in accordance to [#544 (comment)](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/544#issuecomment-715607901). Please let me know if you agree, and I will do that right in this branch.

@atamazov  Please go ahead and make the necessary changes

**Comment by atamazov** [comment]
@JehandadKhan @DrizztDoUrden All done, please have a look.

**Comment by JehandadKhan** [comment]
@atamazov I am the author so I cant approve, otherwise the changes look fine to me.

**Comment by atamazov** [comment]
@JehandadKhan Thanks! Let give @DrizztDoUrden some time to look (while CI is running and I am testing it locally).

**Review by atamazov — APPROVED** [review]
Approving some changes made by @JehandadKhan. Other changes approved on behalf of @JehandadKhan.

**Comment by atamazov** [comment]
@JehandadKhan Testing shows that the 3x3U-only implementation has some performance degradation on certain configs, so I returned back to your original design. The difference is that I always skip non-dynamic solutions and those that "take_long_time_to_build_and_have_narrow_coverage". Also I made the same changes in Find.

Ready for review now.

**Comment by atamazov** [comment]
:cyclone: Local performance testing done, no regressions.

**Comment by atamazov** [comment]
@DrizztDoUrden Please review, thanks.

**Comment by JehandadKhan** [comment]
@atamazov @DrizztDoUrden can you please review this PR  ?

**Review by atamazov — APPROVED** [review]
looks good ;)

**Review by DrizztDoUrden — APPROVED** [review]

**Comment by atamazov** [comment]
All CI tests passed, ready for merging.

**Pull request merged**